File: retropreference/src/main/java/com/twiceyuan/retropreference/typeHandler/BaseTypeHandler.java
package com.twiceyuan.retropreference.typeHandler;

import android.content.SharedPreferences;

/**
 * Created by twiceYuan on 20/01/2017.
 *
 * Base type preferences handler.
 */
public abstract class BaseTypeHandler<Type> {

    protected SharedPreferences mPreferences;

    public BaseTypeHandler(SharedPreferences preferences) {
        mPreferences = preferences;
    }

    public abstract void setValue(String key, Type value);

    public abstract Type get(String key, Type defaultValue);

    public Type defaultValue() {
        return null;
    }

    public void clear(String key) {
        mPreferences.edit().remove(key).apply();
    }
}

File: retropreference/src/main/java/com/twiceyuan/retropreference/typeHandler/BaseTypeHandler.kt
package com.twiceyuan.retropreference.typeHandler

import android.content.SharedPreferences

/**
 * Created by twiceYuan on 20/01/2017.

 * Base type preferences handler.
 */
abstract class BaseTypeHandler<Type>(internal var mPreferences: SharedPreferences) {

    abstract fun setValue(key: String, value: Type?)

    @Throws(ClassCastException::class)
    abstract operator fun get(key: String, defaultValue: Type?): Type?

    open fun defaultValue(): Type? {
        return null
    }

    fun clear(key: String) {
        mPreferences.edit().remove(key).apply()
    }
}
